network/p2p/dns/cache.go

```diff
@@ -0,0 +1,88 @@
+package dns
+
+import (
+	"net"
+	"sync"
+	"time"
+)
+
```

Comment:
Summary and conclusion of all Simon's previous comments on this PR :-)

I thumbs up this PR but welcome the following changes:
* A code comment is placed here to explain this DNS cache implementation.
* The algorithm is changed to a more pragmatic algorithm, better for performance.
* The locking is changed to better performing read write lock.

The code comment suggestion:

Since this is open source and anybody -- especially those without DNS knowledge -- can see this code and make certain (possibly incorrect) assumptions about it, then ideally there would be a code comment here explaining the implementation strategy, and that this is not a usual DNS cache implementation, and designed as a pragmatic implementation especially for flow-go use, and not general purpose use, e.g.

```
// - DNS cache implementation notes:
//   - Generic / possibly expected functionality NOT implemented:
//     - Caches domains for TTL seconds as given by upstream DNS resolver, e.g. [1].
//     - Possibly pre-expire cached domains so no connection time resolve delay.
//   - Actual / pragmatic functionality implemented below:
//     - Caches domains for global (not individual domain record TTL) TTL seconds.
//     - Cached IP is returned even if cached entry expired; so no connection time resolve delay.
//     - Detecting expired cached domain triggers async DNS lookup to refresh cached entry.
// [1] https://en.wikipedia.org/wiki/Name_server#Caching_name_server
```

A more pragmatic algorithm:

Regarding the suggestions for a better performing and/or more correctly (in terms of DNS caching) implemented algorithm. Many of them require significant extra work, and it's not clear how to get hold of the real TTLs per domain. The quickest and dirtiest -- which was independently suggested by @simonhf and @Kay-Zee -- is to use a global TTL for all cached domains, always return the cached IP without delay, but if it has secretly expired then update the cached domain asynchronously in the background. This decouples the CPU used to refresh / lookup the domain from the current action of making a new network connection. It only works if the domain IPs rarely change. This functionality is also described in this future issue [1].

Better performing read write lock:

This functionality is also described in this future issue [1].

[1] https://github.com/dapperlabs/flow-go/issues/5799